### DISCUSSION TITLE
RISC-V: Support for vendor extensions

### DISCUSSION THREAD

**Discussion by gamelaster** [NationalSecurityAgency/ghidra#5374] [Q&A]
Many chips currently use T-Head cores (C906, E907, C910 etc.), whose contains custom RISC-V ISA extensions.
In this particular case, T-Head published it's specs [here](https://github.com/T-head-Semi/thead-extension-spec).

I would like to ask, what is recommended and best way of adding support for those instructions into Ghidra?
@florolf (I hope you don't mind the ping!) made support for some of them via hacky solution, directly modifying `.sinc` files ([the changes are here](https://github.com/NationalSecurityAgency/ghidra/compare/master...florolf:ghidra:t-head)), although I am not sure if this would be possible to get into upstream, without better support of vendor extensions in Ghidra.

Thanks

**Comment by mumbel** [comment]
This is not the way to go about it.  Creating a new sinc file with the extension and then including it in the right order.  In this case, include it before riscv.custom.sinc is included would be the best generic way.  Possibly mix in context register logic and ifdefs.

The straight forward way would likely be a new T-Head slaspec file and include all the other sinc and vendor extension sinc files.  If the riscv.custom.sinc is entirely unused it could be omitted that from that slaspec.   Then add a new entry in the ldefs that is T-Head.

**Reply by madushan1000** [reply]
> The straight forward way would likely be a new T-Head slaspec file and include all the other sinc and vendor extension sinc files

Can I do this in a ghidra plugin?

**Reply by thixotropist** [reply]
Take a look at PR #5778 for an example of how you might add the T-head extensions as sinc files.  That PR handles disassembly for many of the non-vendor-specific RISCV instruction extensions.  Generating good pcode for most of these instructions is another matter.  Personally, I would like to see Ghidra's disassembler handle any of the instruction extensions recognized and tested by the development head of binutils.  It looks like there may be about 100 T-head instruction extensions and a few ventana extensions.

**Reply by madushan1000** [reply]
I was thinking of packaging some of these instructions as ghidra plugin because at least one of the chips I'm working with(bl808), has un-rattified rv extensions(rvv 0.7 and packed simd), so I'm not sure ghidra really wants to merge them. I've never written a ghidra plugin before so I'm not sure if it's possible to do.

**Reply by thixotropist** [reply]
Processor sinc files are essentially plugins themselves - you don't need to recompile all of Ghidra to add or change them.  The java or python plugin framework is probably not what you want for instruction set customization.

It would be great if the Ghidra developers gave their perspective on how they want to handle vendor extensions for RISCV as well as PowerPC and maybe Xtensa.

I expect that a community-sourced set of RISCV extensions might look like this:

* ISA extensions are contained in files within `Ghidra/Processors/RISCV/data/languages`
* all extensions reference public materials - not vendor proprietary materials
* all extensions are factored to show what's ratified, what's unratified but with unique opcode encoding, and what's unratified with possibly overlapping vendor encodings.
* extensions are identified in Ghidra the same way they are in gcc, clang, binutils, elf, and [riscv-profiles](https://github.com/riscv/riscv-profiles/blob/main/profiles.adoc)
* users can pick the processor variant during import, which implies new language entries in `riscv.ldefs` and `riscv.opinion`
* extension pcode exists for all instruction extensions, if only placeholder function calls.  

For bl808 and the rvv 0.7 vector extensions you have an added problem - these are unratified and likely deprecated in favor of the ratified rvv 1.0 vector extensions.  The Ghidra's existing riscv vector support is likely much closer to rvv 0.7 than the rvv 1.0 support contained in PR #5778.

I'll see if I can add some of the Thead extensions to PR #5778, at least those extensions found in the development tip of binutil's gas testsuite.

**Reply by madushan1000** [reply]
thank you for the information :) , I will take a deeper look into adding custom sinc files

**Reply by madushan1000** [reply]
@thixotropist btw if you're going to add t-head instructions to you pr, some of these instructions are already implemented in the ghidra fork mentioned above (https://github.com/florolf/ghidra/commits/t-head)

**Reply by thixotropist** [reply]
Thanks!  The fork you mentioned does a nice job of handling the pcode.  I'll add something similar into my work.

**Reply by thixotropist** [reply]
I've added THead RISCV vendor extensions to PR #5778.  Comments and corrections are appreciated.

Notes:

*  Ghidra disassembly testing uses `binutil-2.41` gas testsuite files as source input and the associated assembler and objdump toolchain to generate reference disassembly.
* Generated Ghidra pcode has *not* been tested and likely contains errors, especially where zero- and sign-extension is called for.

**Reply by madushan1000** [reply]
I just gave it a try, I don't see anything obviously broken compared to the fork I mentioned, at least in my projects. In fact, de-compiler output is much better in some cases.
